File: smalivm/src/main/java/org/cf/smalivm/context/MethodState.java
package org.cf.smalivm.context;

import gnu.trove.set.TIntSet;
import gnu.trove.set.hash.TIntHashSet;

import org.cf.util.ClassNameUtils;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;

public class MethodState extends BaseState {

    public static final int ResultRegister = -1;
    public static final int ReturnRegister = -2;
    public static final int ReturnAddressRegister = -3;
    public static final int ThrowRegister = -4;
    public static final int ExceptionRegister = -5;

    public static final String MUTABLE_PARAMETER_HEAP = "mutable";
    public static final String METHOD_HEAP = "method";
    private static final Logger log = LoggerFactory.getLogger(MethodState.class.getSimpleName());
    private final int parameterCount;
    private final int parameterSize;
    private final TIntSet mutableParameters;

    public MethodState(ExecutionContext context, int registerCount) {
        this(context, registerCount, 0, 0);
    }

    public MethodState(ExecutionContext context, int registerCount, int parameterCount, int parameterSize) {
        super(context, registerCount);

        this.parameterCount = parameterCount;
        this.parameterSize = parameterSize;
        mutableParameters = new TIntHashSet(parameterCount);
    }

    MethodState(MethodState other, ExecutionContext context) {
        super(other, context);

        this.parameterCount = other.parameterCount;
        this.parameterSize = other.parameterSize;
        mutableParameters = new TIntHashSet(other.mutableParameters);
    }

    private MethodState(MethodState parent, ExecutionContext context, TIntSet mutableParameters) {
        super(parent, context);

        this.parameterCount = parent.parameterCount;
        this.parameterSize = parent.parameterSize;
        this.mutableParameters = parent.mutableParameters;
    }

    public void assignExceptionRegister(Throwable exception) {
        String className = ClassNameUtils.toInternal(exception.getClass());
        HeapItem item = new HeapItem(exception, className);
        pokeRegister(ExceptionRegister, item, METHOD_HEAP);
    }

    public void assignParameter(int parameterRegister, HeapItem item) {
        assignRegister(parameterRegister, item, METHOD_HEAP);

        boolean mutable = !item.isImmutable();
        if (mutable) {
            pokeRegister(parameterRegister, item, MUTABLE_PARAMETER_HEAP);
            mutableParameters.add(parameterRegister);
        }
    }

    public void assignParameter(int parameterRegister, Object value, String type) {
        assignParameter(parameterRegister, new HeapItem(value, type));
    }

    public void assignRegister(int register, HeapItem item) {
        assignRegister(register, item, METHOD_HEAP);
    }

    public void assignRegister(int register, Object value, String type) {
        assignRegister(register, new HeapItem(value, type));
    }

    /**
     * Identical to {@link #assignRegister(int, HeapItem)} but also updates any register with an identical value to that
     * stored in the target register with the new value.
     */
    public void assignRegisterAndUpdateIdentities(int register, HeapItem item) {
        assignRegisterAndUpdateIdentities(register, item, METHOD_HEAP);
    }

    public void assignResultRegister(HeapItem item) {
        assignRegister(ResultRegister, item, METHOD_HEAP);
    }

    public void assignResultRegister(Object value, String type) {
        assignRegister(ResultRegister, new HeapItem(value, type));
    }

    public void assignReturnRegister(HeapItem item) {
        pokeRegister(ReturnRegister, item, METHOD_HEAP);
    }

    public void assignReturnRegister(Object value, String type) {
        pokeRegister(ReturnRegister, new HeapItem(value, type), METHOD_HEAP);
    }

    public void assignThrowRegister(HeapItem item) {
        pokeRegister(ThrowRegister, item, METHOD_HEAP);
    }

    public int getParameterCount() {
        return parameterCount;
    }

    public int getParameterStart() {
        return getRegisterCount() - parameterSize;
    }

    @Override
    public MethodState getParent() {
        return (MethodState) super.getParent();
    }

    public MethodLocation getPseudoInstructionReturnInstruction() {
        return (MethodLocation) peekRegister(ReturnAddressRegister).getValue();
    }

    public HeapItem peekExceptionRegister() {
        return peekRegister(ExceptionRegister);
    }

    public HeapItem peekParameter(int parameterRegister) {
        HeapItem item;
        if (mutableParameters.contains(parameterRegister)) {
            item = peekRegister(parameterRegister, MUTABLE_PARAMETER_HEAP);
        } else {
            item = peekRegister(parameterRegister);
        }

        return item;
    }

    public
    @Nullable
    HeapItem peekRegister(int register) {
        if (register == MethodState.ResultRegister) {
            if (!hasRegister(register, METHOD_HEAP)) {
                if (getParent() != null && !getParent().hasRegister(register, METHOD_HEAP)) {
                    // ResultRegister can only be read by the instruction immediately after it's set.
                    // It's not in this instruction or its parent, so it effectively doesn't exist.
                    // log.warn("Attempting to read result register but it's not in current or parent context!
                    // Returning null.");
                    return null;
                }
            }
        }

        return peekRegister(register, METHOD_HEAP);
    }

    public HeapItem peekResultRegister() {
        return peekRegister(ResultRegister);
    }

    public HeapItem peekReturnRegister() {
        return peekRegister(ReturnRegister);
    }

    public HeapItem peekThrowRegister() {
        return peekRegister(ThrowRegister);
    }

    public void pokeRegister(int register, HeapItem item) {
        pokeRegister(register, item, METHOD_HEAP);
    }

    public void pokeRegister(int register, Object value, String type) {
        pokeRegister(register, new HeapItem(value, type));
    }

    public HeapItem readRegister(int register) {
        return readRegister(register, METHOD_HEAP);
    }

    public HeapItem readResultRegister() {
        HeapItem item = readRegister(ResultRegister, METHOD_HEAP);

        return item;
    }

    public HeapItem readReturnRegister() {
        return readRegister(ReturnRegister);
    }

    public void setPseudoInstructionReturnLocation(MethodLocation location) {
        // Pseudo instructions like array-data-payload need return addresses.
        pokeRegister(ReturnAddressRegister, location, METHOD_HEAP);
    }

    @Override
    public String toString() {
        return toString(true);
    }

    public String toString(boolean onlyPeekCachedRegisters) {
        StringBuilder sb = new StringBuilder();
        int localsCount = getRegisterCount() - getParameterCount();
        sb.append("params: ").append(parameterCount).append(", ");
        sb.append("locals: ").append(localsCount).append('\n');
        StringBuilder ctx = new StringBuilder();
        for (int register = 0; register < getRegisterCount(); register++) {
            if (onlyPeekCachedRegisters && !hasRegister(register)) {
                continue;
            }

            ctx.append('v').append(register);
            boolean isLocal = register < getParameterStart();
            if (!isLocal) {
                ctx.append("(p").append(register - getParameterStart()).append(')');
            }
            HeapItem item = peekRegister(register);
            ctx.append(": ").append(item).append('\n');

            register += Utils.getRegisterSize(item.getType());
            if (Utils.getRegisterSize(item.getType()) == 2) {
                register += 1;
            }
        }

        if (onlyPeekCachedRegisters && hasRegister(ResultRegister)) {
            ctx.append("result: ").append(peekRegister(ResultRegister)).append('\n');
        }

        if (onlyPeekCachedRegisters && hasRegister(ReturnRegister)) {
            ctx.append("return: ").append(peekRegister(ReturnRegister)).append('\n');
        }

        if (ctx.length() > 0) {
            ctx.setLength(ctx.length() - 1);
        }
        sb.append(ctx);

        return sb.toString();
    }

    public boolean wasRegisterRead(int register) {
        return wasRegisterRead(register, METHOD_HEAP);
    }

    MethodState getChild(ExecutionContext childContext) {
        MethodState child = new MethodState(this, childContext, mutableParameters);

        return child;
    }

    boolean hasRegister(int register) {
        return hasRegister(register, METHOD_HEAP);
    }

}

File: smalivm/src/main/java/org/cf/smalivm/context/MethodState.kt
package org.cf.smalivm.context

import org.cf.smalivm.context.BaseState
import gnu.trove.set.TIntSet
import kotlin.jvm.JvmOverloads
import org.cf.smalivm.context.ExecutionContext
import gnu.trove.set.hash.TIntHashSet
import org.cf.smalivm.context.MethodState
import org.cf.util.ClassNameUtils
import org.cf.smalivm.context.HeapItem
import org.cf.util.Utils
import org.jf.dexlib2.builder.MethodLocation
import org.slf4j.LoggerFactory
import java.lang.StringBuilder

class MethodState : BaseState {
    val parameterCount: Int
    private val parameterSize: Int
    private val mutableParameters: TIntSet

    @JvmOverloads
    constructor(context: ExecutionContext, registerCount: Int, parameterCount: Int = 0, parameterSize: Int = 0) : super(context, registerCount) {
        this.parameterCount = parameterCount
        this.parameterSize = parameterSize
        mutableParameters = TIntHashSet(parameterCount)
    }

    internal constructor(other: MethodState, context: ExecutionContext) : super(other, context) {
        parameterCount = other.parameterCount
        parameterSize = other.parameterSize
        mutableParameters = TIntHashSet(other.mutableParameters)
    }

    private constructor(parent: MethodState, context: ExecutionContext, mutableParameters: TIntSet) : super(parent, context) {
        parameterCount = parent.parameterCount
        parameterSize = parent.parameterSize
        this.mutableParameters = parent.mutableParameters
    }

    fun assignExceptionRegister(exception: Throwable) {
        val className = ClassNameUtils.toInternal(exception.javaClass)
        val item = HeapItem(exception, className)
        pokeRegister(ExceptionRegister, item, METHOD_HEAP)
    }

    fun assignParameter(parameterRegister: Int, item: HeapItem) {
        assignRegister(parameterRegister, item, METHOD_HEAP)
        val mutable = !item.isImmutable
        if (mutable) {
            pokeRegister(parameterRegister, item, MUTABLE_PARAMETER_HEAP)
            mutableParameters.add(parameterRegister)
        }
    }

    fun assignParameter(parameterRegister: Int, value: Any?, type: String?) {
        assignParameter(parameterRegister, HeapItem(value, type))
    }

    fun assignRegister(register: Int, item: HeapItem?) {
        assignRegister(register, item, METHOD_HEAP)
    }

    fun assignRegister(register: Int, value: Any?, type: String?) {
        assignRegister(register, HeapItem(value, type))
    }

    /**
     * Identical to [.assignRegister] but also updates any register with an identical value to that
     * stored in the target register with the new value.
     */
    fun assignRegisterAndUpdateIdentities(register: Int, item: HeapItem) {
        assignRegisterAndUpdateIdentities(register, item, METHOD_HEAP)
    }

    fun assignResultRegister(item: HeapItem?) {
        assignRegister(ResultRegister, item, METHOD_HEAP)
    }

    fun assignResultRegister(value: Any?, type: String?) {
        assignRegister(ResultRegister, HeapItem(value, type))
    }

    fun assignReturnRegister(item: HeapItem) {
        pokeRegister(ReturnRegister, item, METHOD_HEAP)
    }

    fun assignReturnRegister(value: Any?, type: String) {
        pokeRegister(ReturnRegister, HeapItem(value, type), METHOD_HEAP)
    }

    fun assignThrowRegister(item: HeapItem) {
        pokeRegister(ThrowRegister, item, METHOD_HEAP)
    }

    val parameterStart: Int
        get() = registerCount - parameterSize

    fun getParent(): MethodState {
        return super.parent as MethodState
    }

    val pseudoInstructionReturnInstruction: MethodLocation?
        get() = peekRegister(ReturnAddressRegister)!!.value as MethodLocation?

    fun peekParameter(parameterRegister: Int): HeapItem? {
        val item: HeapItem?
        item = if (mutableParameters.contains(parameterRegister)) {
            peekRegister(parameterRegister, MUTABLE_PARAMETER_HEAP)
        } else {
            peekRegister(parameterRegister)
        }
        return item
    }

    fun peekRegister(register: Int): HeapItem? {
        if (register == ResultRegister) {
            if (!hasRegister(register, METHOD_HEAP)) {
                if (parent != null && !getParent().hasRegister(register, METHOD_HEAP)) {
                    // ResultRegister can only be read by the instruction immediately after it's set.
                    // It's not in this instruction or its parent, so it effectively doesn't exist.
                    // log.warn("Attempting to read result register but it's not in current or parent context!
                    // Returning null.");
                    return null
                }
            }
        }
        return peekRegister(register, METHOD_HEAP)
    }

    fun peekExceptionRegister(): HeapItem? {
        return peekRegister(ExceptionRegister)
    }

    fun peekResultRegister(): HeapItem? {
        return peekRegister(ResultRegister)
    }

    fun peekReturnRegister(): HeapItem? {
        return peekRegister(ReturnRegister)
    }

    fun peekThrowRegister(): HeapItem? {
        return peekRegister(ThrowRegister)
    }

    fun pokeRegister(register: Int, item: HeapItem?) {
        pokeRegister(register, item, METHOD_HEAP)
    }

    fun pokeRegister(register: Int, value: Any?, type: String?) {
        pokeRegister(register, HeapItem(value, type))
    }

    fun readRegister(register: Int): HeapItem {
        return readRegister(register, METHOD_HEAP)
    }

    fun readResultRegister(): HeapItem {
        return readRegister(ResultRegister, METHOD_HEAP)
    }

    fun readReturnRegister(): HeapItem {
        return readRegister(ReturnRegister)
    }

    fun setPseudoInstructionReturnLocation(location: MethodLocation?) {
        // Pseudo instructions like array-data-payload need return addresses.
        pokeRegister(ReturnAddressRegister, location, METHOD_HEAP)
    }

    override fun toString(): String {
        return toString(true)
    }

    fun toString(onlyPeekCachedRegisters: Boolean): String {
        val sb = StringBuilder()
        val localsCount = registerCount - parameterCount
        sb.append("params: ").append(parameterCount).append(", ")
        sb.append("locals: ").append(localsCount).append('\n')
        val ctx = StringBuilder()
        var register = 0
        while (register < registerCount) {
            if (onlyPeekCachedRegisters && !hasRegister(register)) {
                register += 1
                continue
            }
            ctx.append('v').append(register)
            val isLocal = register < parameterStart
            if (!isLocal) {
                ctx.append("(p").append(register - parameterStart).append(')')
            }
            val item = peekRegister(register)
            ctx.append(": ").append(item).append('\n')
            val registerSize = if (item == null) 1 else Utils.getRegisterSize(item.type)
            register += registerSize
        }
        if (hasRegister(ResultRegister)) {
            ctx.append("result: ").append(peekRegister(ResultRegister)).append('\n')
        }
        if (hasRegister(ReturnRegister)) {
            ctx.append("return: ").append(peekRegister(ReturnRegister)).append('\n')
        }
        if (ctx.length > 0) {
            ctx.setLength(ctx.length - 1)
        }
        sb.append(ctx)
        return sb.toString()
    }

    fun wasRegisterRead(register: Int): Boolean {
        return wasRegisterRead(register, METHOD_HEAP)
    }

    fun getChild(childContext: ExecutionContext): MethodState {
        return MethodState(this, childContext, mutableParameters)
    }

    fun hasRegister(register: Int): Boolean {
        return hasRegister(register, METHOD_HEAP)
    }

    companion object {
        const val ResultRegister = -1
        const val ReturnRegister = -2
        const val ReturnAddressRegister = -3
        const val ThrowRegister = -4
        const val ExceptionRegister = -5
        const val MUTABLE_PARAMETER_HEAP = "mutable"
        const val METHOD_HEAP = "method"
        private val log = LoggerFactory.getLogger(MethodState::class.java.simpleName)
    }
}
